src/dbt_osmosis/core/column_level_knowledge_propagator.py

```diff
@@ -84,3 +87,51 @@ def get_node_columns_with_inherited_knowledge(
         family_tree = _build_node_ancestor_tree(manifest, node)
         knowledge = _inherit_column_level_knowledge(manifest, family_tree, placeholders)
         return knowledge
+
+    @staticmethod
+    def update_undocumented_columns_with_prior_knowledge(
+        undocumented_columns: Iterable[str],
+        node: ManifestNode,
+        yaml_file_model_section: Dict[str, Any],
+        knowledge: Knowledge,
+        skip_add_tags: bool,
+        skip_merge_meta: bool,
+        add_progenitor_to_meta: bool,
+    ) -> int:
+        """Update undocumented columns with prior knowledge in node and model simultaneously
+        THIS MUTATES THE NODE AND MODEL OBJECTS so that state is always accurate"""
+        inheritables = ["description"]
+        if not skip_add_tags:
+            inheritables.append("tags")
+        if not skip_merge_meta:
+            inheritables.append("meta")
+
+        changes_committed = 0
+        for column in undocumented_columns:
+            prior_knowledge: ColumnLevelKnowledge = get_prior_knowledge(knowledge, column)
+            progenitor = prior_knowledge.pop("progenitor", None)
+            prior_knowledge = {k: v for k, v in prior_knowledge.items() if k in inheritables}
+            if add_progenitor_to_meta and progenitor:
+                prior_knowledge.setdefault("meta", {})
+                prior_knowledge["meta"]["osmosis_progenitor"] = progenitor
+            if not prior_knowledge:
+                continue
+            if column not in node.columns:
+                node.columns[column] = ColumnInfo.from_dict({"name": column, **prior_knowledge})
+            else:
+                node.columns[column] = ColumnInfo.from_dict(
```

Comment:
The original implementation was `model_column.update(prior_knowledge)`. While writing the test, I realized that this code was not updating as intended, so I fixed it. The original code had the behavior that `_extra` in `ColumnInfo` would return a modified object (the original object was not modified).

like this ↓

```python
ColumnInfo(name='first_name', description="Customer's first name. PII.", meta={}, data_type=None, constraints=[], quote=None, tags=[], _extra={'kwargs': {'name': 'first_name', 'meta': {'osmosis_progenitor': 'model.jaffle_shop_duckdb.stg_customers'}}})
```

- ref:
  - https://github.com/z3z1ma/dbt-osmosis/blob/main/src/dbt_osmosis/core/osmosis.py#L1039
  - https://github.com/dbt-labs/dbt-core/blob/d91265411061463b42d6f4d6d94e094a42a96d2b/core/dbt/contracts/graph/nodes.py#L219
  - https://github.com/dbt-labs/dbt-core/blob/d91265411061463b42d6f4d6d94e094a42a96d2b/core/dbt/contracts/util.py#L43-L45

Reply:
I see, good to know about that prop 👀